### ISSUE TITLE
Linker errors with Boost and -static triplets

### ISSUE THREAD

**Issue by chrullrich** [Microsoft/vcpkg#483]
A project using Boost does not link when a -static triplet is used. This has two reasons, both related to boost's autolink support (in boost/config/auto_link.hpp) and library renaming.

- The name of any static library is expected to be prefixed with "lib"
- Currently, static Boost implies static CRT, and static-CRT libraries' names must contain the "s" flag

Both these elements are removed by library renaming in boost/portfile.cmake .

A crude fix is to condition all renaming on (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic), i. e. move the lines into the condition directly above. This appears to fix MSBuild. A better fix may be to `#define BOOST_ALL_NO_LIB` in user.hpp, i.e. disable Boost's autolink feature entirely. Because vcpkg always and indiscriminately adds `*.lib` to the linker command line, all Boost libraries are found anyway.

As for CMake, I cannot say what needs to be done because I have not been able to get a CMake project to work; `find_package(Boost)` only works if I define VCPKG_PREFIX_PATH and include the triplet directory. AFAICT, this is not intended to be necessary.

**Comment by codicodi** [comment]
IMHO preserving auto linking capabilities of Boost would be a good thing because it makes it easy to use/redistribute package outside of vcpkg (for example in custom buildsystems).

**Comment by chrullrich** [comment]
Consider that vcpkg already has to keep two build systems happy, MSBuild and CMake. That is difficult enough without catering to an implied third one. Also, using Boost's support for configuring its behavior while maintaining pristine headers at the same time is clearly not possible (and the port currently defines BOOST_ALL_DYN_LINK for non -static triplets, so that ship has sailed anyway).

There are some alternatives to modifying user.hpp that I can think of:

- The other one I mentioned above, where the output from the static build is not renamed. This works with MSBuild, but may not with CMake, and I don't know how to test it (I can build other vcpkg libraries with dependencies on Boost, but I don't know if that is representative for the behavior of projects external to vcpkg).
- A per-project property sheet that defines BOOST_ALL_NO_LIB instead of adding it to user.hpp, but you would have to remember adding it to each project configuration, all the time, every time. This is even more intrusive than user.hpp because it affects the project itself, rather than the libraries.
- Adding the definition to vcpkg.targets. Ugh. Also, CMake.
- If vcpkg can be detected during build, something like this (in user.hpp):

       #ifdef VCPKG_INSTALL
       #define BOOST_ALL_NO_LIB
       #endif

Of all these options, and considering that redistributing the vcpkg-built package directory would require manually deleting one line from one file, I'm in favor of my idea as long as it doesn't break something in either vcpkg or vcpkg's direct clients. Or you could add something to the port that avoids adding the line conditional on some command line option, then use that to build a redistribution package.

Or vcpkg could stop doing that `*.lib` thing, but I don't see how it could, because it would break all libraries that _don't_ have autolinking built in.

**Comment by dcrawford1** [comment]
I just successfully built a cmake project that uses vcpkg boost with find_package(Boost).  I removed all of the library renaming functions in the portfile so the static libraries are installed just as Boost intended.  I had to set the following before including find_package(Boost) in my project's CMakeLists.txt
```
set(Boost_USE_STATIC_LIBS ON)
set(Boost_USE_STATIC_RUNTIME ON)
```
I also disabled auto linking in my CMakeLists.txt
```
add_definitions(-DBOOST_ALL_NO_LIB)
```
I did not have to define VCPKG_PREFIX_PATH or BOOST_ROOT or BOOST_INCLUDEDIR

**Comment by codicodi** [comment]
@dcrawford1 not renaming the libs would certainly fix some issues but then another problem arises: how would another package that depends on boost find it? Maybe we could provide a wrapper around FindBoost that sets `Boost_USE_STATIC_LIBS` and `Boost_USE_STATIC_RUNTIME ` correctly, according to current triplet.

**Comment by dcrawford1** [comment]
Creating a wrapper is not a bad idea.  But, if the consumer is a cmake project which calls find_package(Boost) we could also add a global option to the portfile cmake build step for the "static" triplet case:
```
vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    OPTIONS -DBoost_USE_STATIC_LIBS=ON
)
```
In general, I think we should avoid renaming libraries as much as possible.

**Comment by alexkaratarakis** [comment]
The reasoning behind the current rename behavior:
`vcpkg` only allows one instance of each library to be installed at a time. Therefore, the names of the lib files should not matter; the "capability" of boost is either there or it is not. 

Furthermore, a particular triplet can be configured to build some libraries statically and some dynamically. For example, you may opt to create your own triplet that builds zlib as a dll, and then all other dependencies are built as static libs. A portfile of another library can't easily know what configuration the other libraries were built with.

Therefore, it can't know whether boost is static or dynamic in order to pass the appropriate flag into FindBOOST. We tried stripping out the build-specific identifiers from the names, but it did not work out (https://github.com/Microsoft/vcpkg/issues/369). Therefore, we rename the static names to the dynamic names, and then always look for the dynamic names. Therefore, you always use FindBOOST in the same way and the lib file does the right thing.

**Comment by dcrawford1** [comment]
Well, the problem for me at least was that I could not get my project to work with FindBoost with the renamed libraries.  If I do not define Boost_USE_STATIC_LIBS=ON then FindBoost does find the renamed libraries and it does successfully link one of my programs that just includes boost::chrono.  But, I have another program that links with a lot of boost libraries using the FindBoost imported targets and this one fails on the link with `cannot open file 'libboost_system-vc140-mt-sgd-1_62.lib'`  I am not sure how I get this link error.  This is the original filename, but vcpkg has renamed all of these libraries.  The vcxproj does not contain any references to libboost_system-vc140-mt-sgd-1_62.lib.  I have the commit which defines BOOST_ALL_NO_LIB to disable auto linking.  The FindBoost.cmake file mentions something about implicitly including dependent imported targets, maybe that's it?  Anyway, if I change the vcpkg portfile to not rename libraries and define Boost_USE_STATIC_LIBS=ON then everything works.  

I now understand the rationale for renaming libraries and it makes sense.  I will try to put together a simple test case that demonstrates this problem.

**Comment by dcrawford1** [comment]
Nevermind, I have everything successfully linking now with the renamed boost libraries.  I had another vcpkg library that was linking with the original named static boost libraries.

**Comment by JonesOliver** [comment]
I've managed to link to boost with a sample test project using both the x86-windows and x86-windows-static triplets, but cannot seem to link the boost libraries to an existing project.  Has anyone tried to do this and been successful?  Anyone also had problem using vcpkg, boost and old VS projects? 

I'm simply using the boost::filesystem and can successfully link to a set of static boost libraries that I've built   and maintain myself.

For reference and if anyone finds this useful, to link to the static libs I copied in the following in my vs project file under <PropertyGroup Label="Globals">:

    <VcpkgTriplet>x86-windows-static</VcpkgTriplet>
    <VcpkgEnabled>true</VcpkgEnabled>

but also had to change the code generation setting "Runtime Library" from Multi-threaded Debug DLL /MDd to Multii-threaded Debug /MTd
